shaders/oit_linked_lists/combine.frag

```diff
@@ -0,0 +1,112 @@
+#version 450
+/* Copyright (c) 2023, Google
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 the "License";
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	 http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#define LINKED_LIST_END_SENTINEL	0xFFFFFFFFU
+#define SORTED_FRAGMENT_MAX_COUNT	16U
+
+////////////////////////////////////////
+
+layout(set = 0, binding = 0) uniform SceneConstants
+{
+	mat4 projection;
+	mat4 view;
+	uint sortFragments;
+	uint filtered_blending;
+	uint fragmentMaxCount;
+	uint unused;
+} sceneConstants;
+
+layout(set = 0, binding = 2, r32ui) uniform uimage2D linkedListHeadTex;
+
+layout(set = 0, binding = 3) buffer FragmentBuffer {
+	uvec3 data[];
+} fragmentBuffer;
+
+layout(set = 0, binding = 4) buffer FragmentCounter {
+	uint value;
+} fragmentCounter;
+
+////////////////////////////////////////
+
+layout (location = 0) out vec4 outFragColor;
+
+////////////////////////////////////////
+
+void main()
+{
+	// Reset the atomic counter for the next frame
+	// Note that we don't care about atomicity here, as all threads will write the same value
+	fragmentCounter.value = 0;
+
+	// Get the first fragment index in the linked list
+	uint fragmentIndex = imageLoad(linkedListHeadTex, ivec2(gl_FragCoord.xy)).r;
+	// Reset the list head for the next frame
+	imageStore(linkedListHeadTex, ivec2(gl_FragCoord.xy), uvec4(LINKED_LIST_END_SENTINEL, 0, 0, 0));
+
+	// Copy the fragments into local memory for sorting
+	uvec2 sortedFragments[SORTED_FRAGMENT_MAX_COUNT];
+	uint fragmentCount = 0U;
+	while(fragmentIndex != LINKED_LIST_END_SENTINEL && fragmentCount < SORTED_FRAGMENT_MAX_COUNT)
```

Comment:
You just consider SORTED_FRAGMENT_MAX_COUNT layers on a fragment and ignore any further ones. That way, depending on rendering order and camera position, you might even miss the nearest one.
Maybe you could, if there are more than SORTED_FRAGMENT_MAX_COUNT layers, mix-in the farthest layers in arbitrary order and only have the nearest SORTED_FRAGMENT_MAX_COUNT layers sorted?
That way, it might be even worth to make SORTED_FRAGMENT_MAX_COUNT a UI-controlled parameter.

Reply:
Ah, very good point.

Technically, it should be possible to do a coarse sort first to limit the issue, but it would stray away from the purpose of the sample.
I plan to write a multipass sample, with a completely different technique, that removes the issue.

I added an explanation about the performance/correctness trade-off in the README.
I also added an option to control the maximum number of sorted fragments per pixel, so that users can visualize the artifacts when that number is low.

Reply:
You don't need to do a coarse sort first. Instead of sorting the first encountered fragments, you could as well keep them sorted:
```
	while ( fragmentIndex != LINKED_LIST_END_SENTINEL )
	{
		const uvec3 fragment = fragmentBuffer.data[fragmentIndex];
		fragmentIndex = fragment.x;
		if ( sceneConstants.sortFragments == 1U)
		{
			if ( fragmentCount < sortedFragmentMaxCount )
			{
				uint i=fragmentCount;
				for ( ; (0<i) && (fragment.z < sortedFragments[i-1].y) ; --i)
				{
					sortedFragments[i] = sortedFragments[i-1];
				}
				sortedFragments[i] = fragment.yz;
				++fragmentCount;
			}
			else if ( sortedFragments[0].y <= fragment.z )
			{
				color = blend( color, sortedFragments[0].x);
				uint i=0;
				for ( ; (i<sortedFragmentMaxCount-1) && (sortedFragments[i+1].y <= fragment.z) ; ++i)
				{
					sortedFragments[i] = sortedFragments[i+1];
				}
				sortedFragments[i] = fragment.yz;
			}
			else
			{
				color = blend( color, fragment.y);
			}
		}
		else
		{
			sortedFragments[fragmentCount] = fragment.yz;
			++fragmentCount;
		}
	}
```
Granted, a bit more complex, but you don't need to sort anymore.
When you skip the blending operation in this sequence, you would nicely see the increasing depth when you increase the sortedFragmentMaxCount.

Reply:
Ah, I see!
I misunderstood you original comment.
Let me try it.

Reply:
I used your algorithm for the combine pass. It works very well!
I can barely see any artifacts even with 4-5 sorted fragments.

I assume it might get worse with lower transparency, but it is way better than just ignoring the fragments.
Please have a look and thanks for the trick!

Reply:
Looks great!
Glad, I could help here.